include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h

```diff
@@ -1548,7 +1424,7 @@ __serial_merge(const _Rng1& __rng1, const _Rng2& __rng2, _Rng3& __rng3, _Index1
     }
     else
     {
-        const auto __n = ::std::min<_Index1>(__n1 + __n2 - __start3, __chunk);
```

Comment:
Actually, the lines 1533 `if (__start1 >= __n1)`,  1541 `else if (__start2 >= __n2)`, 1560 `if(++__start2 == __n2)` and 1574 `if (++__start1 == __n1)` guarantee that the buffers cannot be overrun.

Reply:
In the previous two if branches, they are protected from overflow by `::std::min` when assigning `__n`, but here in the `else` branch, the conditionals you mention just detect the case where we no longer have to compare, and can just write the rest of the `__chunk` from a single list.  

It does not protect the overflow case for the other list or the output.  Inside those conditionals, it still uses `__n` as the limit for the loop which if assigned to `__chunk` may be able to overflow the ranges involved.

Reply:
why "it does not protect"? could you please give me an example?
(__start1  ? __start2 ? __n1 ? __n2 ?)

In the previous branch, the check `const auto __n = ::std::min<_Index1>(__n1 + __n2 - __start3, __chunk);` was redundant...

Reply:
`__chunk=4`
`__n1 = 1`
`__n2 = 1`
`__start1 = __start2 = __start3 = 0`

this skips the `if` and `else if` to get to the `else` branch of `__serial_merge`.  
now 
`__n = 4` (with the PR as written)

Lets say that the result of `__comp(__val2, __val1)` is `true` (but it doesn't matter).

Now we assign into `__rng3[0]= __rng2[0]`.
Then we hit `if (++__start2 == __n2)` which is true, so we enter the branch, where we iterate for `i=1`, `2`, `3`, for all of those cases `i < __n` because `__n == chunk ==4`.
this assigns `_rng3[1] = __rng1[0]`  which is fine.
Then it overflows in reading and writing with 
`_rng3[2] = __rng1[1]` and `_rng3[3] = __rng1[2]`.

Similar overflow occurs if `__comp(__val2, __val1)` was `false`.

Reply:
oops... yes, I missed such possible cases.. I'll take a look.

Reply:
fixed

Reply:
fixed